torch_glow/src/FuseLinear.cpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "FuseLinear.h"
+
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/alias_analysis.h>
+#include <torch/csrc/jit/passes/common_subexpression_elimination.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/subgraph_rewrite.h>
+
+namespace glow {
+
+/// This is mainly copied from pytorch/tvm
+/// This pass fuse the addmm or matmul + add generated by JIT back to linear
```

Comment:
Maybe dumb question, but: if I understand correctly, we're essentially recreating the FC (linear) from the mm + add, and then importing it as linear, correct? Why can't we just import it as linear and then do this fusion in Glow?

Reply:
@jfix71 no it's not a dumb question, when glow's jit ir pass sees this, it's lowered to addmm, matmul, etc but would be nice to have it as linear and it's easy to fuse it here. Eventually the hope is this will move to pytorch directly then we won't have to worry about this, user will use a linear op and glow will see a linear op. Glow could introduce a folding pass that does this too.

Reply:
Ok got it. NBD to just go with this as is since it's already done, but my two cents are that we'll probably want to support importing addmm or matmul + add by themselves (if they aren't already supported). And that it's also useful to have this folding to FC logic in Glow too, since it would help all frontend importers. Whereas this seems like a temporary workaround that we'll throw away eventually and only helps PyTorch.

Reply:
100% agree with all of that